### PR TITLE
Various changes in Python API

### PR DESCRIPTION
**Pull request by paulromano**
This pull request makes a few changes/additions:

- `Universe.plot` now has two extra arguments, `colors` (dictionary mapping cells/materials to RGB(A) tuples) and `filename` (for saving plot to file instead of display).
- Most `__deepcopy__` methods have been removed since they don't do anything special. I'm holding off on making more radical changes like autogenerating new IDs for a deepcopy since we don't have consensus on it yet.
- A new `openmc.model` subpackage was added with a `TRISO` class that helps ease modeling TRISO fuel particles. There's also a `create_triso_lattice` function which is loosly based on @smharper's one-off script for placing TRISO particles in a lattice, duplicating them where they overlap lattice boundaries. You can look at the new test_triso to see how the class and function are used.

@wbinventor Note that I've held off on adding a test for the Python plotting capabilities mostly due to the fact that there's no good way to hash a PNG (and writing raw RGBA data from matplotlib was [broken](https://github.com/matplotlib/matplotlib/pull/6519) apparently). The other reason I'm deferring is that we should really decide what we want to do with respect to unit testing.

### CODE REVIEW

**Comment by wbinventor** [comment]
I will review this later today, though it might be nice if @smharper can help out, particularly with regards to `openmc.model`. Thanks for holding off on the ID-related changes to `__deepcopy__`, we can cogitate on that together a bit more first. 

As for the testing your new plotting methods, I feel your pain with matplotlib. I'm not sure how much you looked into running diffs on matplotlib-generated figures, but it is worse than you might imagine. The images generated on different machines with the exact same install of matplotlib (and the exact same `rcparams` settings) may generate figures of slightly different width x height specs, even if you try to specify them manually. Needless to say, this makes a diff between a test and reference figure difficult. 

All of our tests for the plotting module in OpenMOC by generating an RGB histogram of the pixel in a test/reference plot and then diff'ing the histogram (as recommended on Stack Overflow) as illustrated [here](https://github.com/mit-crpg/OpenMOC/blob/develop/tests/testing_harness.py#L346). This hacky solution is known to miss small but perhaps meaningful deviations in figures ([link])(https://github.com/mit-crpg/OpenMOC/issues/304), but will at least catch any medium to large errors in the figure.

File: openmc/arithmetic.py [hunk 1]
```diff
@@ -532,23 +474,6 @@ def __eq__(self, other):
     def __ne__(self, other):
         return not self == other
 
-    def __deepcopy__(self, memo):
-        existing = memo.get(id(self))
```
Comment (wbinventor) [inline openmc/arithmetic.py]: I'm curious - is this line included in the default implementation? Seeing this is now sparking my memory of *one* of the reasons I went about implementing `__deepcopy__` methods back in the day - to boost performance by eliminating redundant copies of objects in CSG trees (*i.e.*, this line of code). I want to say that this is not included in the default implementation, but I'm not sure. Regardless, I don't *think* it matters for the `Cross*` and `Aggregate*` classes anyway.

Comment (wbinventor) [inline openmc/arithmetic.py]: I also may be wrong, the `memo` may be used in this way by default.

File: openmc/filter.py [hunk 2]
```diff
@@ -104,27 +104,6 @@ def __lt__(self, other):
...
         return hash(repr(self))
 
-    def __deepcopy__(self, memo):
-        existing = memo.get(id(self))
-
-        # If this is the first time we have tried to copy this object, create a copy
-        if existing is None:
-            clone = type(self).__new__(type(self))
-            clone._type = self.type
-            clone._bins = copy.deepcopy(self.bins, memo)
```
Comment (wbinventor) [inline openmc/filter.py]: Now in this case one *would* want a deep rather than shallow copy of the bins. If I understand correctly, the default implementation will actually deep copy all of the class attributes?

File: openmc/arithmetic.py (hunk 1 above)
Comment (paulromano) [inline openmc/arithmetic.py]: Yes, as far as I can tell `memo` is used this way by default. Here's proof:
```python
In [1]: import openmc, copy

In [2]: c = openmc.Cell()

In [3]: l = [c, c]

In [4]: l_copy = copy.deepcopy(l)

In [5]: [id(x) for x in l_copy]
Out[5]: [140129480509648, 140129480509648]

In [6]: memo = {id(c): 'See Will, I told you so'}

In [7]: l_copy2 = copy.deepcopy(l, memo)

In [8]: l_copy2
Out[8]: ['See Will, I told you so', 'See Will, I told you so']
```

File: openmc/filter.py (hunk 2 above)
Comment (paulromano) [inline openmc/filter.py]: Yes, deepcopy operates recursively.

Comment (wbinventor) [inline openmc/filter.py]: Awesome :-) I'm probably remembering the early days when I implemented deep
copies without memo and they were very slow. Good to know that isn't a
problem for us here.

File: tests/test_triso/test_triso.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        # Create TRISO particles
+        materials = [fuel, porous_carbon, ipyc, sic, opyc]
+        radii = np.array([212.5, 312.5, 347.5, 382.5, 422.5])*1e-4
+        trisos = []
+        random.seed(1)
+        for i in range(100):
+            # Randomly sample location
+            x = random.uniform(-0.5, 0.5)
+            y = random.uniform(-0.5, 0.5)
+            z = random.uniform(-0.5, 0.5)
```
Comment (smharper) [inline tests/test_triso/test_triso.py]: This can sample particles that are split by the reflective boundary condition.  These should be changed to something like
```python
            x = random.uniform(-0.5+radii[-1], 0.5-radii[-1])
```
Or you could move the reflective BCs outwards by `radii[-1]`

File: tests/test_triso/test_triso.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        max_y = openmc.YPlane(y0=0.5, boundary_type='reflective')
+        min_z = openmc.ZPlane(z0=-0.5, boundary_type='reflective')
+        max_z = openmc.ZPlane(z0=0.5, boundary_type='reflective')
+        box = openmc.Cell(region=+min_x & -max_x & +min_y & -max_y & +min_z & -max_z)
+
+        # Create lattice
+        ll, ur = box.region.bounding_box
+        shape = (3, 3, 3)
+        lattice = openmc.model.create_triso_lattice(
+            trisos, ll, (ur - ll)/shape, shape, graphite)
```
Comment (smharper) [inline tests/test_triso/test_triso.py]: Bearing in mind that our test suite is a big part of our documentation, I would recommend making it more clear that this is a lattice pitch
```python
        pitch = (ur - ll) / shape
        lattice = openmc.model.create_triso_lattice(
            trisos, ll, pitch, shape, graphite)
```

File: openmc/model/triso.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        if lattice.ndim == 2:
+            (i_min, j_min), p = lattice.find_element(ll)
+            (i_max, j_max), p = lattice.find_element(ur)
+            return list(np.broadcast(*np.ogrid[
+                j_min:j_max+1, i_min:i_max+1]))
+        else:
+            (i_min, j_min, k_min), p = lattice.find_element(ll)
+            (i_max, j_max, k_max), p = lattice.find_element(ur)
+            return list(np.broadcast(*np.ogrid[
+                k_min:k_max+1, j_min:j_max+1, i_min:i_max+1]))
```
Comment (smharper) [inline openmc/model/triso.py]: There should be a check here to make sure that all the `*_min` and `*_max` indices are valid.  Unfortunately, we can't use `lattice.is_valid_index` because `lattice._universes` isn't allocated yet

File: openmc/model/triso.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            if idx in sorted(triso_locations):
+                # Create copy of TRISO particle with materials preserved and
+                # different cell/surface IDs
+                t_copy = copy.deepcopy(t)
+                for c, c_copy in zip(t.cells, t_copy.cells):
+                    c_copy.id = None
+                    c_copy.fill = c.fill
+                for s in t_copy._surfaces:
+                    s.id = None
+                triso_locations[idx].append(t_copy)
```
Comment (smharper) [inline openmc/model/triso.py]: Do we actually need a `shape` argument?  I think we could get rid of it with something like
```python
    triso_locations = {}
    for t in trisos:
        for idx_ in t.classify(lattice):
            idx = tuple(int(x) for x in idx_)
            # Create copy of TRISO particle with materials preserved and
            # different cell/surface IDs
            t_copy = copy.deepcopy(t)
            for c, c_copy in zip(t.cells, t_copy.cells):
                c_copy.id = None
                c_copy.fill = c.fill
            for s in t_copy._surfaces:
                s.id = None
            triso_locations.setdefault(idx, [])
            triso_locations[idx].append(t_copy)
```
Otherwise, we should check to make sure the given shape actually fits all the trisos.  Similarly, we need to check the lower_left.  If you use an algorithm like the one above, an improper lower_left will give a `triso_locations` dictionary containing keys with negative values.

File: openmc/model/triso.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+    lattice = openmc.RectLattice()
+    lattice.lower_left = lower_left
+    lattice.pitch = pitch
+
+    indices = list(np.broadcast(*np.ogrid[:shape[2], :shape[1], :shape[0]]))
+    triso_locations = {idx: [] for idx in indices}
+    for t in trisos:
+        for idx in t.classify(lattice):
+            if idx in sorted(triso_locations):
```
Comment (smharper) [inline openmc/model/triso.py]: To be clear, there should never be a case where
```python
idx in sorted(triso_locations)
```
is False.  It is possible for the `TRISO.classify` method to return indices which do not actually contain the TRISO, but they still need to be valid indices for the lattice.  Otherwise, TRISOs will get cut by the lattice boundary.

**Comment by smharper** [comment]
@paulromano, this implementation has all of the TRISO surfaces on the same universe level.  I would really recommend using a two-level approach where the upper universe only has one surface and cell per TRISO.  That cell is filled with a lower universe containing the rest of the surfaces.

There are two big advantages to this method:  The `distance_to_boundary` calculation only has to check one surface per TRISO at that coordinate level.  It also removes redundant surfaces from the geometry.xml which will significantly reduce the time required to read geometry.xml (reading the geometry can easily take several minutes).

I have a rough example implementation [in my repo](https://github.com/smharper/openmc/tree/various-pyapi-changes).  The test you added runs twice as many neutrons/s with these changes.  I suspect the speedup will be even larger for higher packing fractions.

File: openmc/model/triso.py (hunk 6 above)
Comment (paulromano) [inline openmc/model/triso.py]: My thought is that a user will probably know the bounds for the lattice, so without pre-assigning universes (which is not possible), the only way to get that information across is by specifying the shape. After thinking about this, I'm leaning towards raising a warning if an idx returned by classify is not found in triso_locations. So it'll look something like
```python
    indices = list(np.broadcast(*np.ogrid[:shape[2], :shape[1], :shape[0]]))
    triso_locations = {idx: [] for idx in indices}
    for t in trisos:
        for idx in t.classify(lattice):
            if idx in sorted(triso_locations):
                # Create copy of TRISO particle with materials preserved and                                                                                                        
                # different cell/surface IDs                                                                                                                                        
                t_copy = copy.deepcopy(t)
                for c, c_copy in zip(t.cells, t_copy.cells):
                    c_copy.id = None
                    c_copy.fill = c.fill
                for s in t_copy._surfaces:
                    s.id = None
                triso_locations[idx].append(t_copy)
            else:
                warnings.warn('TRISO particle is partially or completely '
                              'outside of lattice.', UserWarning)
```

What do you think?

**Comment by paulromano** [comment]
@smharper Nice idea with the multiple universes! Though to be clear, even with the current scheme there should still only be one surface per TRISO that is checked during `distance_to_boundary`. That being said, there is certainly an input reduction and performance advantage to having a single universe for many TRISOs though, so I will adopt your idea.

**Comment by smharper** [comment]
Hmm, I see now that you're right about `distance_to_boundary`.  I wonder what causes the 2x performance difference.  I doubt it would be `find_cell`.  I'm tempted to profile the runtime.

**Comment by paulromano** [comment]
Well, since the single universe (and cells within) for TRISOs gets reused as a particle is tracking through, it should be a little more cache-friendly all around.

**Comment by smharper** [comment]
I did profile the code.  It turns out that `find_cell` really is a bottleneck.  With the 1-level approach, `find_cell` takes 67% of the runtime.  With 2-level, it takes 37%.  I bet this is because of the call from `cross_lattice` since that call doesn't use a reduced search list.

**Comment by paulromano** [comment]
@smharper Your comments have been addressed. I also added a commit fixing an issue with `Material.add_nuclide` whereby if you add two nuclides with the same name but different xs, the first one gets overwritten.

File: openmc/model/triso.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+import openmc
+import openmc.checkvalue as cv
+
+class TRISO(object):
+    """Tristructural-isotopic (TRISO) micro fuel particle
+
+    Parameters
+    ----------
+    outer_radius : int
```
Comment (smharper) [inline openmc/model/triso.py]: `int -> Real`

**Comment by smharper** [comment]
I'm happy.  @wbinventor, have all of your concerns been addressed?

**Comment by wbinventor** [comment]
This ship is ready to sail!

**Comment by paulromano** [comment]
Thanks guys! Can we also sail my one-liner in #672?

**Comment by smharper** [comment]
Shesh, you give 'em an inch...
